### ISSUE TITLE
Ignore conditional block when condition is CancellationToken.IsCancellationRequested

### ISSUE THREAD

**Issue by skarllot** [zompinc/sync-method-generator#61]
If the asynchronous method has a conditional block testing if CancellationToken is canceled that block should be ignored.

```csharp
public ValueTask Test(CancellationToken cancellationToken)
{
    if (cancellationToken.IsCancellationRequested)
    {
        // ...
        return ValueTask.FromCanceled(cancellationToken);
    }
    // ...
}
```

**Comment by lsoft** [comment]
@skarllot not sure that ISG should care about this issue. Imagine more complex if statements like `if (cancellationToken.IsCancellationRequested && ...)`.  removing conditional subtree is not an option here, ISG will have to rewrite source code in a complex way like replacing `cancellationToken.IsCancellationRequested` with `false`. Not sure ISG should perform such complex rewriting, let's wait for @virzak resolution.

I guess this case is the case for `#if SYNC_ONLY`.

**Comment by skarllot** [comment]
On second thought I believe you are right, I prefer using compiler conditional.

**Comment by virzak** [comment]
Great issue.

The source generator has already picked an aggressive/advanced way of dealing with such issues. The goal is to make is as close to the code one would write by hand.

For example if it sees `IProgress<int>` used like this:

```cs
if (true)
{
    progress.Report(1);
}
```

the entire if block will be dropped.

This approach did land us in a bit of a hot water, for example in https://github.com/zompinc/sync-method-generator/issues/45#issuecomment-1893956960

I think that the example above should be dropped without forcing the user to invoke `SYNC_ONLY`, just because of how common that scenario can be.

When combined with `&&` it should be dropped as well.
When negated (`!`) the body should get unrolled.

If there's `while` instead of `if`, the block should be dropped for `ct.IsCancellationRequested` and be made while (true) for `!ct.IsCancellationRequested`

Looking at code [occurrences](https://github.com/search?q=IsCancellationRequested+language%3AC%23+org%3Adotnet&type=code) under `dotnet` org, the most common uses are:

- `while (!ct.IsCancellationRequested)`
- `if (ct.IsCancellationRequested)`

**Comment by virzak** [comment]
This fix is available in 1.3.35. Let me know how this works out.

At the end not unrolling the if block because you could have a duplicate variable:

```cs
if (true)
{
    var myVar = 1;
}

if (true)
{
    var myVar = 1;
}
```

Taking either myVar out of scope would cause a compilation issue.
